File: python/tests/test_feedback_analyser.py
"""test_feedback_analyser.py

Exercises FeedbackAnalyser (persona-adaptation deltas from a window of signals)
and the InMemoryFeedbackStore. Mirrors the TypeScript feedback_analyser.test.ts
and the C# FeedbackAnalyser rules / FeedbackStoreTests.
"""
from __future__ import annotations

import struct
import uuid
from datetime import datetime, timedelta, timezone

import pytest

from circle_ai.memory.feedback_analyser import FeedbackAnalyser, PersonaAdaptation
from circle_ai.memory.feedback_signal import FeedbackPolarity, FeedbackSignal
from circle_ai.memory.stores import InMemoryFeedbackStore


def _f32(x: float) -> float:
    return struct.unpack("<f", struct.pack("<f", x))[0]


# FP32-narrowed deltas — must equal the C# `float` literals exactly.
VERBOSITY_DOWN = _f32(-0.1)
VERBOSITY_UP = _f32(0.05)

_BASE = datetime(2023, 11, 14, tzinfo=timezone.utc)
_seq = 0


def make(
    polarity: FeedbackPolarity,
    at: datetime | None = None,
    user: str = "user",
) -> FeedbackSignal:
    global _seq
    ts = at if at is not None else _BASE + timedelta(seconds=_seq)
    _seq += 1
    return FeedbackSignal(
        id=uuid.uuid4(),
        recorded_at_utc=ts,
        user_text=user,
        assistant_text="response",
        polarity=polarity,
    )


# ══════════════════════════════════════════════════════════════════════════
# FeedbackAnalyser
# ══════════════════════════════════════════════════════════════════════════


def test_rejects_a_window_size_below_1():
    with pytest.raises(ValueError):
        FeedbackAnalyser(0)


def test_returns_zero_deltas_for_empty_signal_set():
    a = FeedbackAnalyser().analyse([])
    assert a.verbosity_delta == 0
    assert a.formality_delta == 0
    assert a.preferred_topics == []


def test_drops_verbosity_when_over_70_percent_negative():
    analyser = FeedbackAnalyser()
    # 8 negative + 2 positive = 80% negative.
    signals = [make(FeedbackPolarity.NEGATIVE) for _ in range(8)]
    signals += [make(FeedbackPolarity.POSITIVE) for _ in range(2)]

    a = analyser.analyse(signals)
    assert a.verbosity_delta == VERBOSITY_DOWN
    assert a.formality_delta == 0
    assert a.preferred_topics == []


def test_raises_verbosity_when_over_70_percent_positive():
    analyser = FeedbackAnalyser()
    signals = [make(FeedbackPolarity.POSITIVE) for _ in range(8)]
    signals += [make(FeedbackPolarity.NEGATIVE) for _ in range(2)]

    a = analyser.analyse(signals)
    assert a.verbosity_delta == VERBOSITY_UP


def test_leaves_verbosity_at_0_for_balanced_window():
    analyser = FeedbackAnalyser()
    signals = [make(FeedbackPolarity.POSITIVE) for _ in range(5)]
    signals += [make(FeedbackPolarity.NEGATIVE) for _ in range(5)]

    assert analyser.analyse(signals).verbosity_delta == 0


def test_treats_exactly_70_percent_as_not_crossing_threshold():
    analyser = FeedbackAnalyser(10)
    # Exactly 7/10 negative — 0.70 is not > 0.70.
    signals = [make(FeedbackPolarity.NEGATIVE) for _ in range(7)]
    signals += [make(FeedbackPolarity.POSITIVE) for _ in range(3)]

    assert analyser.analyse(signals).verbosity_delta == 0


def test_only_considers_most_recent_window_size_signals_newest_first():
    analyser = FeedbackAnalyser(3)
    # Older bulk is positive; the 3 newest are negative → window is 100% negative.
    older = [
        make(FeedbackPolarity.POSITIVE, datetime(2023, 1, 1, tzinfo=timezone.utc) + timedelta(seconds=i))
        for i in range(10)
    ]
    newest = [
        make(FeedbackPolarity.NEGATIVE, datetime(2024, 1, 1, tzinfo=timezone.utc) + timedelta(seconds=i))
        for i in range(3)
    ]

    a = analyser.analyse([*older, *newest])
    assert a.verbosity_delta == VERBOSITY_DOWN


def test_ignores_correction_signals_in_the_ratio():
    analyser = FeedbackAnalyser()
    # 8 negative + 2 correction = 8/10 = 80% negative → down.
    signals = [make(FeedbackPolarity.NEGATIVE) for _ in range(8)]
    signals += [make(FeedbackPolarity.CORRECTION) for _ in range(2)]
    assert analyser.analyse(signals).verbosity_delta == VERBOSITY_DOWN


def test_rejects_a_null_signal_set():
    with pytest.raises(ValueError):
        FeedbackAnalyser().analyse(None)  # type: ignore[arg-type]


def test_persona_adaptation_is_a_dataclass():
    a = PersonaAdaptation(0.0, 0.0, [])
    assert a.verbosity_delta == 0.0
    assert a.formality_delta == 0.0
    assert a.preferred_topics == []


# ══════════════════════════════════════════════════════════════════════════
# InMemoryFeedbackStore
# ══════════════════════════════════════════════════════════════════════════


async def test_store_rejects_a_null_signal():
    store = InMemoryFeedbackStore()
    with pytest.raises(ValueError):
        await store.add_async(None)  # type: ignore[arg-type]


async def test_store_add_increments_the_count():
    store = InMemoryFeedbackStore()
    await store.add_async(make(FeedbackPolarity.POSITIVE))
    assert await store.count_async() == 1


async def test_store_get_recent_on_empty_returns_empty():
    store = InMemoryFeedbackStore()
    assert await store.get_recent_async(10) == []


async def test_store_get_recent_returns_newest_first():
    store = InMemoryFeedbackStore()
    now = datetime.now(timezone.utc)
    await store.add_async(make(FeedbackPolarity.POSITIVE, now - timedelta(minutes=10), "old"))
    await store.add_async(make(FeedbackPolarity.NEGATIVE, now, "new"))

    result = await store.get_recent_async(10)
    assert len(result) == 2
    assert result[0].user_text == "new"


async def test_store_positive_ratio_returns_none_with_no_signals():
    store = InMemoryFeedbackStore()
    assert await store.positive_ratio_async() is None


async def test_store_positive_ratio_returns_1_when_all_positive():
    store = InMemoryFeedbackStore()
    await store.add_async(make(FeedbackPolarity.POSITIVE))
    await store.add_async(make(FeedbackPolarity.POSITIVE))
    assert await store.positive_ratio_async() == 1.0


async def test_store_positive_ratio_returns_right_fraction_for_mixed():
    store = InMemoryFeedbackStore()
    await store.add_async(make(FeedbackPolarity.POSITIVE))
    await store.add_async(make(FeedbackPolarity.POSITIVE))
    await store.add_async(make(FeedbackPolarity.NEGATIVE))
    ratio = await store.positive_ratio_async()
    assert ratio is not None and 0.66 < ratio < 0.68  # 2/3


async def test_store_evicts_oldest_when_max_signals_exceeded_fifo():
    store = InMemoryFeedbackStore(3)
    for i in range(5):
        await store.add_async(make(FeedbackPolarity.POSITIVE, None, f"u{i}"))
    assert await store.count_async() == 3


def test_store_rejects_non_positive_max_signals():
    with pytest.raises(ValueError):
        InMemoryFeedbackStore(0)

File: c/tests/test_feedback_analyser.c
/*
 * test_feedback_analyser.c — FeedbackAnalyser + InMemoryFeedbackStore (C11).
 *
 * Mirrors the verified TypeScript feedback_analyser.test.ts and the C#
 * FeedbackAnalyser rules. The FP32 delta constants (-0.1f, +0.05f) are asserted
 * exactly.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>
#include "circle_ai/circle_ai.h"

/* monotonic default timestamps so ordering is deterministic per call */
static int64_t g_seq = 0;

static ca_feedback_signal_rec_t make(ca_feedback_polarity_t pol, int64_t at_ms, const char *user) {
    ca_feedback_signal_rec_t s = {0};
    s.id = NULL;
    s.recorded_at_ms = at_ms >= 0 ? at_ms : (1700000000000LL + (g_seq++) * 1000);
    s.user_text = (char *)user;      /* borrowed — analyse copies internally */
    s.assistant_text = (char *)"response";
    s.polarity = pol;
    return s;
}

int main(void) {
    /* ── FeedbackAnalyser ── */

    /* rejects a window size below 1 */
    assert(ca_feedback_analyser_create(0) == NULL);

    ca_feedback_analyser_t *a20 = ca_feedback_analyser_create(20);
    assert(a20);

    /* empty signal set → zero deltas */
    {
        ca_persona_adaptation_t out;
        ca_feedback_analyser_analyse(a20, NULL, 0, &out);
        assert(out.verbosity_delta == 0.0f);
        assert(out.formality_delta == 0.0f);
        assert(out.topic_count == 0 && out.preferred_topics == NULL);
        ca_persona_adaptation_free(&out);
    }

    /* > 70% negative → -0.1f  (8 neg + 2 pos) */
    {
        ca_feedback_signal_rec_t sig[10];
        for (int i=0;i<8;++i) sig[i] = make(CA_FEEDBACK_POLARITY_NEGATIVE, -1, "user");
        for (int i=8;i<10;++i) sig[i] = make(CA_FEEDBACK_POLARITY_POSITIVE, -1, "user");
        ca_persona_adaptation_t out;
        ca_feedback_analyser_analyse(a20, sig, 10, &out);
        assert(out.verbosity_delta == -0.1f);
        assert(out.formality_delta == 0.0f);
        assert(out.topic_count == 0);
        ca_persona_adaptation_free(&out);
    }

    /* > 70% positive → +0.05f  (8 pos + 2 neg) */
    {
        ca_feedback_signal_rec_t sig[10];
        for (int i=0;i<8;++i) sig[i] = make(CA_FEEDBACK_POLARITY_POSITIVE, -1, "user");
        for (int i=8;i<10;++i) sig[i] = make(CA_FEEDBACK_POLARITY_NEGATIVE, -1, "user");
        ca_persona_adaptation_t out;
        ca_feedback_analyser_analyse(a20, sig, 10, &out);
        assert(out.verbosity_delta == 0.05f);
        ca_persona_adaptation_free(&out);
    }

    /* balanced → 0 (5 pos + 5 neg) */
    {
        ca_feedback_signal_rec_t sig[10];
        for (int i=0;i<5;++i) sig[i] = make(CA_FEEDBACK_POLARITY_POSITIVE, -1, "user");
        for (int i=5;i<10;++i) sig[i] = make(CA_FEEDBACK_POLARITY_NEGATIVE, -1, "user");
        ca_persona_adaptation_t out;
        ca_feedback_analyser_analyse(a20, sig, 10, &out);
        assert(out.verbosity_delta == 0.0f);
        ca_persona_adaptation_free(&out);
    }

    /* exactly 70% is NOT > 70% (strict) — 7/10 negative with window 10 */
    {
        ca_feedback_analyser_t *a10 = ca_feedback_analyser_create(10);
        ca_feedback_signal_rec_t sig[10];
        for (int i=0;i<7;++i) sig[i] = make(CA_FEEDBACK_POLARITY_NEGATIVE, -1, "user");
        for (int i=7;i<10;++i) sig[i] = make(CA_FEEDBACK_POLARITY_POSITIVE, -1, "user");
        ca_persona_adaptation_t out;
        ca_feedback_analyser_analyse(a10, sig, 10, &out);
        assert(out.verbosity_delta == 0.0f);
        ca_persona_adaptation_free(&out);
        ca_feedback_analyser_destroy(a10);
    }

    /* only the most-recent windowSize (newest-first): older bulk positive, 3
     * newest negative → window 3 is 100% negative → down */
    {
        ca_feedback_analyser_t *a3 = ca_feedback_analyser_create(3);
        ca_feedback_signal_rec_t sig[13];
        for (int i=0;i<10;++i) sig[i] = make(CA_FEEDBACK_POLARITY_POSITIVE, 1000 + i, "user");
        for (int i=0;i<3;++i)  sig[10+i] = make(CA_FEEDBACK_POLARITY_NEGATIVE, 9000000 + i, "user");
        ca_persona_adaptation_t out;
        ca_feedback_analyser_analyse(a3, sig, 13, &out);
        assert(out.verbosity_delta == -0.1f);
        ca_persona_adaptation_free(&out);
        ca_feedback_analyser_destroy(a3);
    }

    /* Correction signals ignored in the ratio: 8 neg + 2 correction = 8/10 → down */
    {
        ca_feedback_signal_rec_t sig[10];
        for (int i=0;i<8;++i) sig[i] = make(CA_FEEDBACK_POLARITY_NEGATIVE, -1, "user");
        for (int i=8;i<10;++i) sig[i] = make(CA_FEEDBACK_POLARITY_CORRECTION, -1, "user");
        ca_persona_adaptation_t out;
        ca_feedback_analyser_analyse(a20, sig, 10, &out);
        assert(out.verbosity_delta == -0.1f);
        ca_persona_adaptation_free(&out);
    }

    ca_feedback_analyser_destroy(a20);

    /* ── InMemoryFeedbackStore ── */

    /* rejects a non-positive maxSignals */
    assert(ca_feedback_store_create(0) == NULL);

    /* rejects a null signal */
    {
        ca_feedback_store_t *store = ca_feedback_store_create(100);
        assert(!ca_feedback_store_add(store, NULL));
        ca_feedback_store_destroy(store);
    }

    /* add increments the count */
    {
        ca_feedback_store_t *store = ca_feedback_store_create(100);
        ca_feedback_signal_rec_t s = make(CA_FEEDBACK_POLARITY_POSITIVE, -1, "user");
        assert(ca_feedback_store_add(store, &s));
        assert(ca_feedback_store_count(store) == 1);
        ca_feedback_store_destroy(store);
    }

    /* getRecent on an empty store returns empty */
    {
        ca_feedback_store_t *store = ca_feedback_store_create(100);
        size_t n;
        ca_feedback_signal_rec_t *r = ca_feedback_store_get_recent(store, 10, &n);
        assert(n == 0 && r == NULL);
        ca_feedback_store_destroy(store);
    }

    /* getRecent returns newest-first */
    {
        ca_feedback_store_t *store = ca_feedback_store_create(100);
        ca_feedback_signal_rec_t older = make(CA_FEEDBACK_POLARITY_POSITIVE, 1000, "old");
        ca_feedback_signal_rec_t newer = make(CA_FEEDBACK_POLARITY_NEGATIVE, 2000, "new");
        assert(ca_feedback_store_add(store, &older));
        assert(ca_feedback_store_add(store, &newer));
        size_t n;
        ca_feedback_signal_rec_t *r = ca_feedback_store_get_recent(store, 10, &n);
        assert(n == 2);
        assert(strcmp(r[0].user_text, "new") == 0);
        ca_feedback_signal_free_array(r, n);
        ca_feedback_store_destroy(store);
    }

    /* positiveRatio: null when empty */
    {
        ca_feedback_store_t *store = ca_feedback_store_create(100);
        double ratio;
        assert(!ca_feedback_store_positive_ratio(store, &ratio));
        ca_feedback_store_destroy(store);
    }

    /* positiveRatio: 1.0 when all positive */
    {
        ca_feedback_store_t *store = ca_feedback_store_create(100);
        ca_feedback_signal_rec_t s1 = make(CA_FEEDBACK_POLARITY_POSITIVE, -1, "u");
        ca_feedback_signal_rec_t s2 = make(CA_FEEDBACK_POLARITY_POSITIVE, -1, "u");
        ca_feedback_store_add(store, &s1);
        ca_feedback_store_add(store, &s2);
        double ratio;
        assert(ca_feedback_store_positive_ratio(store, &ratio));
        assert(ratio == 1.0);
        ca_feedback_store_destroy(store);
    }

    /* positiveRatio: 2/3 for mixed */
    {
        ca_feedback_store_t *store = ca_feedback_store_create(100);
        ca_feedback_signal_rec_t s1 = make(CA_FEEDBACK_POLARITY_POSITIVE, -1, "u");
        ca_feedback_signal_rec_t s2 = make(CA_FEEDBACK_POLARITY_POSITIVE, -1, "u");
        ca_feedback_signal_rec_t s3 = make(CA_FEEDBACK_POLARITY_NEGATIVE, -1, "u");
        ca_feedback_store_add(store, &s1);
        ca_feedback_store_add(store, &s2);
        ca_feedback_store_add(store, &s3);
        double ratio;
        assert(ca_feedback_store_positive_ratio(store, &ratio));
        assert(ratio > 0.66 && ratio < 0.68);
        ca_feedback_store_destroy(store);
    }

    /* FIFO eviction when maxSignals exceeded */
    {
        ca_feedback_store_t *store = ca_feedback_store_create(3);
        for (int i=0;i<5;++i) {
            ca_feedback_signal_rec_t s = make(CA_FEEDBACK_POLARITY_POSITIVE, -1, "u");
            ca_feedback_store_add(store, &s);
        }
        assert(ca_feedback_store_count(store) == 3);
        ca_feedback_store_destroy(store);
    }

    /* analyse a store directly */
    {
        ca_feedback_store_t *store = ca_feedback_store_create(100);
        for (int i=0;i<8;++i) { ca_feedback_signal_rec_t s = make(CA_FEEDBACK_POLARITY_NEGATIVE, -1, "u"); ca_feedback_store_add(store, &s); }
        for (int i=0;i<2;++i) { ca_feedback_signal_rec_t s = make(CA_FEEDBACK_POLARITY_POSITIVE, -1, "u"); ca_feedback_store_add(store, &s); }
        ca_feedback_analyser_t *an = ca_feedback_analyser_create(20);
        ca_persona_adaptation_t out;
        ca_feedback_analyser_analyse_store(an, store, &out);
        assert(out.verbosity_delta == -0.1f);
        ca_persona_adaptation_free(&out);
        ca_feedback_analyser_destroy(an);
        ca_feedback_store_destroy(store);
    }

    printf("test_feedback_analyser: all assertions passed\n");
    return 0;
}
